qa/L0_logging/log_format_test.py

```diff
@@ -0,0 +1,271 @@
+os.makedirs(test_logs_directory)
+
+# Regular expression pattern to capture the headers and rows
+table_regex = re.compile(
+    r'\+[-+]+\+\n'  # Match the top border
+    r'\| (?P<header>.*?) \|\n'  # Capture the header
+    r'\+[-+]+\+\n'  # Match the header border
+    r'(?P<rows>(?:\| .*? \|\n)*)'  # Capture the rows
+    r'\+[-+]+\+',  # Match the bottom border
+    re.DOTALL  # Enable dot to match newlines
+)
+
+# Regular expression pattern
+default_pattern = r'(?P<level>\w)(?P<month>\d{2})(?P<day>\d{2}) (?P<timestamp>\d{2}:\d{2}:\d{2}\.\d{6}) (?P<pid>\d+) (?P<file>[\w\.]+):(?P<line>\d+)] (?P<message>.*)'
+
+# Compile the regex pattern
+default_regex = re.compile(default_pattern, re.DOTALL)
+
+LEVELS = set({"E", "W", "I"})
+
+FORMATS = [
+    ("default", default_regex),
+    ("ISO8601", ""),
+    ("default_unescaped", ""),
+    ("ISO8601_unescaped", ""),
+]
+
+IDS = ["default", "ISO8601", "default_unescaped", "ISO8601_unescaped"]
+
+validators = {}
+
+def validator(func):
+    validators[func.__name__.replace('validate_','')] = func
+    return func
+
+@validator
+def validate_level(level):
+    assert level in LEVELS
+
+@validator
+def validate_month(month):
+    assert month.isdigit()
+    month = int(month)
+    assert month >= 1 and month <= 12
+
+@validator
+def validate_day(day):
+    assert day.isdigit()
+    day = int(day)
+    assert day >= 1 and day <= 31
+
+@validator
+def validate_timestamp(timestamp):
+    parse_timestamp(timestamp)
+
+@validator
+def validate_pid(pid):
+    assert pid.isdigit()
+
+@validator
+def validate_file(file_):
+    assert Path(file_).name is not None
+
+@validator
+def validate_line(line):
+    assert line.isdigit()
+
+def validate_table(table):
+    header = table.group("header").strip().split('|')
+    rows = table.group("rows").strip().split('\n')
+
+    # Process each row
+    parsed_rows = []
+    for row in rows:
+        if row:
+            row_data = [r.strip() for r in row.split('|')[1:-1]]
+            parsed_rows.append(row_data)
+   
+    for row in parsed_rows:
+        assert len(row)==len(header)
+    
+@validator
+def validate_message(message):
+    heading, obj = message.split('\n',1)
+    if heading:
+        try:
+            json.loads(heading)
+        except json.JSONDecodeError as e:
+            raise Exception(f"{e} First line of message in log record is not a valid JSON string")
+        except Exception as e:
+            raise type(e)(f"{e} First line of message in log record is not a valid JSON string")
+    if len(obj):
+        obj = obj.strip()       
+        match = table_regex.search(obj)
+        if match:
+            validate_table(match)
+        else:
+            google.protobuf.text_format.Parse(obj,grpcclient.model_config_pb2.ModelConfig())
+
+class TestLogFormat:
+    @pytest.fixture(autouse=True)
+    def setup(self, request):
+        test_case_name = request.node.name
+        self._server_options = {}
+        self._server_options["log-verbose"] = 256
```

Comment:
Is this `256` supposed to be testing something about the `Log::Level` being a `uint8_t`?

Reply:
we don't have a defined max for log verbose - it is currently defined as a uint32_t - though I thought about changing it to a uint8_t.

this is distinct from log level, this the verbosity level

Reply:
I see, just looked like magic number to me so was curious

Reply:
open to other thoughts here - could do max unsigned int -